components/brave_ads/core/internal/user_engagement/site_visit/site_visit.cc

```diff
@@ -37,94 +41,187 @@ void SiteVisit::RemoveObserver(SiteVisitObserver* observer) {
   observers_.RemoveObserver(observer);
 }
 
+///////////////////////////////////////////////////////////////////////////////
+
+bool SiteVisit::IsPageLanding(const int32_t tab_id) const {
+  return base::Contains(page_lands_, tab_id);
+}
+
 void SiteVisit::MaybeLandOnPage(const TabInfo& tab) {
   if (!last_clicked_ad_) {
     // No ad interactions have occurred in the current browsing session.
     return;
   }
 
-  if (tab_id_ == tab.id) {
-    // This tab is being processed.
-    return;
+  if (!IsPageLanding(tab.id)) {
+    CheckIfLandedOnPage(tab, *last_clicked_ad_);
   }
-  tab_id_ = tab.id;
 
-  if (!DomainOrHostExists(tab.redirect_chain, last_clicked_ad_->target_url)) {
+  // Reset the last clicked ad to prevent multiple landings on the same ad.
+  last_clicked_ad_.reset();
+}
+
+void SiteVisit::CheckIfLandedOnPage(const TabInfo& tab, const AdInfo& ad) {
+  if (!DomainOrHostExists(tab.redirect_chain, ad.target_url)) {
     return BLOG(1, "Visited URL does not match the last clicked ad");
   }
 
-  CheckIfLandedOnPage(tab);
+  StartPageLand(tab.id, ad,
+                base::BindRepeating(&SiteVisit::CheckIfLandedOnPageCallback,
+                                    base::Unretained(this), tab));
```

Comment:
why not `weak_factory_.GetWeakPtr()`? you already have it `SiteVisit::LandedOnPage`.

Reply:
So we do. Great catch. Thanks

Reply:
Fixed along with fixing some areas of the codebase which could also be changed.